File: examples/web/src/shell/libraryPick.ts
import type { AppMode } from "../AppShellTypes";

const SEP = "\u001f";

export type LibraryPick =
  | { kind: "unified" }
  | { kind: "all"; mode: AppMode }
  | { kind: "folder"; mode: AppMode; folder: string }
  | { kind: "source"; mode: AppMode; folder: string; sourceId: string };

export type LibraryItem = {
  mode: AppMode;
  id: string;
  title: string;
  publishedAt: number | null;
  url: string;
  thumb?: string;
};

export function pickMode(pick: LibraryPick): AppMode {
  return pick.kind === "unified" ? "news" : pick.mode;
}

export function parentPick(pick: LibraryPick): LibraryPick | null {
  if (pick.kind === "folder") return { kind: "all", mode: pick.mode };
  if (pick.kind === "source") return { kind: "folder", mode: pick.mode, folder: pick.folder };
  return null;
}

export function encodeLibraryPick(pick: LibraryPick): string {
  if (pick.kind === "unified") return "unified";
  if (pick.kind === "all") return `all${SEP}${pick.mode}`;
  if (pick.kind === "folder") return `folder${SEP}${pick.mode}${SEP}${pick.folder}`;
  return `source${SEP}${pick.mode}${SEP}${pick.folder}${SEP}${pick.sourceId}`;
}

export function parseLibraryPick(raw: string | null | undefined): LibraryPick {
  const text = raw?.trim() ?? "";
  if (!text || text === "unified") return { kind: "unified" };
  const parts = text.split(SEP);
  const mode = (parts[1] === "podcast" || parts[1] === "booru" ? parts[1] : "news") as AppMode;
  if (parts[0] === "all") return { kind: "all", mode };
  if (parts[0] === "folder" && parts[2]) return { kind: "folder", mode, folder: parts[2] };
  if (parts[0] === "source" && parts[3]) {
    return { kind: "source", mode, folder: parts[2] ?? "", sourceId: parts.slice(3).join(SEP) };
  }
  return { kind: "all", mode: "news" };
}

export function sortLibraryItems(items: LibraryItem[], oldestFirst: boolean): LibraryItem[] {
  const ordered = [...items].sort((left, right) => {
    const a = left.publishedAt ?? Number.MAX_SAFE_INTEGER;
    const b = right.publishedAt ?? Number.MAX_SAFE_INTEGER;
    return a === b ? left.id.localeCompare(right.id) : a - b;
  });
  return oldestFirst ? ordered : ordered.reverse();
}

export function mergeUnified(
  news: LibraryItem[],
  podcasts: LibraryItem[],
  boards: LibraryItem[],
  oldestFirst: boolean,
): LibraryItem[] {
  return sortLibraryItems([...news, ...podcasts, ...boards], oldestFirst);
}

export function shareUrl(item: LibraryItem | null | undefined): string | undefined {
  const url = item?.url.trim();
  return url || undefined;
}

File: examples/android/app/src/main/java/org/aetherfeed/app/ui/navigation/LibraryPick.kt
package org.aetherfeed.app.ui.navigation

private const val SEP = '\u001f'

sealed class LibraryPick {
    data object Unified : LibraryPick()
    data class All(val mode: AppDestination) : LibraryPick()
    data class Folder(val mode: AppDestination, val folder: String) : LibraryPick()
    data class Source(val mode: AppDestination, val folder: String, val sourceId: String) : LibraryPick()
}

fun LibraryPick.destination(): AppDestination = when (this) {
    LibraryPick.Unified -> AppDestination.News
    is LibraryPick.All -> mode
    is LibraryPick.Folder -> mode
    is LibraryPick.Source -> mode
}

fun LibraryPick.parent(): LibraryPick? = when (this) {
    LibraryPick.Unified -> null
    is LibraryPick.All -> null
    is LibraryPick.Folder -> LibraryPick.All(mode)
    is LibraryPick.Source -> LibraryPick.Folder(mode, folder)
}

fun LibraryPick.expandKeys(): Set<String> = when (this) {
    LibraryPick.Unified -> emptySet()
    is LibraryPick.All -> setOf(mode.name)
    is LibraryPick.Folder -> setOf(mode.name, "${mode.name}:$folder")
    is LibraryPick.Source -> setOf(mode.name, "${mode.name}:$folder")
}

fun encodeLibraryPick(pick: LibraryPick): String = when (pick) {
    LibraryPick.Unified -> "unified"
    is LibraryPick.All -> "all$SEP${pick.mode.name}"
    is LibraryPick.Folder -> "folder$SEP${pick.mode.name}$SEP${pick.folder}"
    is LibraryPick.Source -> "source$SEP${pick.mode.name}$SEP${pick.folder}$SEP${pick.sourceId}"
}

fun parseLibraryPick(raw: String?): LibraryPick {
    val text = raw?.trim().orEmpty()
    if (text.isEmpty() || text == "unified") return LibraryPick.Unified
    val parts = text.split(SEP)
    return when (parts.firstOrNull()) {
        "all" -> LibraryPick.All(parseAppDestination(parts.getOrNull(1)))
        "folder" -> {
            val mode = parseAppDestination(parts.getOrNull(1))
            val folder = parts.getOrNull(2)?.trim().orEmpty()
            if (folder.isEmpty()) LibraryPick.All(mode) else LibraryPick.Folder(mode, folder)
        }
        "source" -> {
            val mode = parseAppDestination(parts.getOrNull(1))
            val folder = parts.getOrNull(2)?.trim().orEmpty()
            val id = parts.drop(3).joinToString(SEP.toString()).trim()
            if (id.isEmpty()) LibraryPick.All(mode) else LibraryPick.Source(mode, folder, id)
        }
        else -> LibraryPick.All(AppDestination.News)
    }
}
